### ISSUE TITLE
ipynb_playground/keras/convmnist/keras_conv_autoencoder_mnist.ipynb needs updated to latest version of keras

### ISSUE THREAD

**Issue by isaacgerg** [julienr/ipynb_playground#1]
This example won't work anymore as keras interface has changed.

Have changes and am willing to send if you'd like.

**Comment by julienr** [comment]
That's right. I would be happy to merge a PR if you have fixed it.

**Comment by isaacgerg** [comment]
I am new to github but use Hg daily.  How do I do this PR  you talk about?

**Comment by isaacgerg** [comment]
Julien,

It will not be a trivial fix I believe.  Your code depends on Convolution2D
having a border_mode of "full" which extends the convolution by the size of
the filter kernel.  This functionality is no longer supported so you can
never grow your layers back to 28 x 28.

Does this sound right?

Isaac

**Comment by julienr** [comment]
From a quick look, I think they only renamed `border_mode='full'` to `border_mode='same'`. In the keras theano backend, they translate same to full :
https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L559

So changing the `border_mode` to same should do it.

Back to your first comment : a "PR" is a github pull request. What you have to do to contribute on github is the following :
- fork this repository (fork button in the upper-right of the page). This will create **isaacgerg/ipynb_playground**
- create a branch on **isaacgerg/ipynb_playground**, let's call it *autoencoder_fix*
- push you changes to this branch
- then you can follow [this guide](https://help.github.com/articles/creating-a-pull-request/) to do a pull request

By the way, your changes don't have to be final to open a pull request. As long as I haven't merged the PR, you can push to your branch and this will update the PR. So if you want to work on this, go ahead and open a PR with the changes you've already made and I can help you get it working.

**Comment by isaacgerg** [comment]
The sizes are no the same with the change from full to same.  Full pads the
image, same does not.  Therefore, you won't get back to the input size.

Isaac

**Comment by julienr** [comment]
That's right... :-(

But it seems keras added a [ZeroPadding2D layer](https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L455). So we should be able to add one of those before every convolution in the decoder (and compute the correct padding).

**Comment by isaacgerg** [comment]
im working on it...

**Comment by julienr** [comment]
Great, let me know if I can help

**Comment by isaacgerg** [comment]
Here are my code changes.   can you cut and paste and see if they work.

    # Layer-wise pretraining
    #
https://github.com/julienr/ipynb_playground/blob/master/keras/convmnist/keras_conv_autoencoder_mnist.ipynb
    # - Build the encoder pipeline
    # TODO idg - fix comments
    encoder = models.Sequential()
    encoder.add(Convolution2D(32, 5, 5, border_mode='valid',
input_shape=(1, 28,28)))
    encoder.add(MaxPooling2D(pool_size=(2, 2)))
    encoder.add(Convolution2D(32, 3, 3, border_mode='valid')) #,
border_mode='valid'), input_shape(1,3,3))
    encoder.add(MaxPooling2D(pool_size=(2, 2)))  # after this, we have 32 *
12 * 12
    encoder.add(Flatten())
    encoder.add(Dense(128))
    encoder.add(Dense(10))
    #encoder.add(Dense(32*5*5))
    #encoder.add(Dense(128))


    # idg
    #optimizer = SGD(lr=0.01, momentum=0.8, decay=0., nesterov=True)
    #model.compile(loss='mean_squared_error', optimizer='adadelta')
    #encoder.compile(loss='mean_squared_error', optimizer=optimizer)
    #encoder.fit(X_train, X_train_orig, batch_size=batch_size, nb_epoch=2,
show_accuracy=True, validation_data=(X_train, X_train_orig))

    #plot(encoder)

    print('-- encoder')
    #print_layer_shapes(encoder, (1, 1, 28, 28))

    #encode = theano.function([encoder.get_input(train=False)],
encoder.get_output(train=False))

    # - Decoder
    decoder = models.Sequential()
    #decoder.add(Dense(10, )))
    decoder.add(Dense(128, input_shape=(10,)))
    decoder.add(Dense(32*14*14))
    decoder.add(Reshape(dims=(32, 14, 14)))
    #decoder.add(UnPooling2D(poolsize=(2, 2)))
    decoder.add(ZeroPadding2D(padding=(2,2)))
    decoder.add(Convolution2D(32, 3, 3, border_mode='same'))
    decoder.add(ZeroPadding2D(padding=(5,5)))
    decoder.add(Convolution2D(1, 5, 5, border_mode='same'))

**Comment by isaacgerg** [comment]
you actually want to use UpSample2d no zeropadding2d.  my mistake.

**Comment by isaacgerg** [comment]
So this is what I've settled on.  However, I dont think the zeropadding2d
is the right model to use.  I will need to think about this more.  Can you
run this with your configuration and see if you get the same graphs?

Thanks,
Isaac

    # - Build the encoder pipeline
    encoder = models.Sequential()
    encoder.add(Convolution2D(32, 5, 5, border_mode='valid',
input_shape=(1, 28,28)))
    encoder.add(MaxPooling2D(pool_size=(2, 2)))
    encoder.add(Convolution2D(32, 3, 3, border_mode='valid'))
    encoder.add(MaxPooling2D(pool_size=(2, 2)))
    encoder.add(Flatten())
    encoder.add(Dense(128))
    encoder.add(Dense(10))

    # For running with fixed weight
    encode = theano.function([encoder.get_input(train=False)],
encoder.get_output(train=False))

    # - Decoder
    decoder = models.Sequential()
    decoder.add(Dense(128, input_shape=(10,)))
    decoder.add(Dense(32*5*5))
    decoder.add(Reshape(dims=(32, 5, 5)))
    decoder.add(UpSampling2D(size=(2,2)))
    decoder.add(Convolution2D(32, 3, 3, border_mode='same'))
    decoder.add(ZeroPadding2D(padding=(1,1)))
    decoder.add(UpSampling2D(size=(2,2)))
    decoder.add(Convolution2D(1, 5, 5, border_mode='same'))
    decoder.add(ZeroPadding2D(padding=(2,2)))

**Comment by julienr** [comment]
Thanks for the code. I think I got it to work. I had to change two things from your snippet :
- Put the ZeroPadding2D before the Convolution2D
- Use my own UnPooling2D layer. For some reason, using keras' UpSampling2D layers creates "blocks" in the reconstructed image. I'm not quite sure why since it looks quite similar to my UnPooling2D layer

It's still running but it looks good so far. I will push tomorrow.

**Comment by isaacgerg** [comment]
I found the same thing as your bullet 2.  I just emailed Francois about it.

**Comment by isaacgerg** [comment]
Can you send me your unpooling2dlayer code?

**Comment by isaacgerg** [comment]
regarding the upsampling issue: bug report here
https://github.com/fchollet/keras/issues/1226

**Comment by julienr** [comment]
@isaacgerg I pushed the updated notebook.

The code for UnPooling2D is included in the notebook, but here it is :
```python
# TODO: For some reason, this doesn't do the exact same thing as UpSampling2D. Not sure why
class UnPooling2D(Layer):
    """A 2D Repeat layer"""
    def __init__(self, poolsize=(2, 2)):
        super(UnPooling2D, self).__init__()
        self.input = T.tensor4()
        self.poolsize = poolsize
        
    @property
    def output_shape(self):
        input_shape = self.input_shape
        return (input_shape[0], input_shape[1],
                self.poolsize[0] * input_shape[2],
                self.poolsize[1] * input_shape[3])

    def get_output(self, train):
        X = self.get_input(train)
        s1 = self.poolsize[0]
        s2 = self.poolsize[1]
        output = X.repeat(s1, axis=2).repeat(s2, axis=3)
        return output

    def get_config(self):
        return {"name":self.__class__.__name__,
            "poolsize":self.poolsize}
```
